src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,909 @@
+  void SetAdjustedFitness(Population& population) {
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      ssize_t speciesSize = population.aSpecies[i].aGenomes.size();
+      for (ssize_t j=0; j<speciesSize; ++j) {
+        double fitness = population.aSpecies[i].aGenomes[j].Fitness();
+        double adjustedFitness = fitness / speciesSize;
+        population.aSpecies[i].aGenomes[j].AdjustedFitness(adjustedFitness);
+      }
+    }
+  }
+
+  // Aggregate population's genomes.
+  void AggregateGenomes(Population& population, std::vector<Genome>& genomes) {
+    genomes.clear();
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      for (ssize_t j=0; j<population.aSpecies[i].aGenomes.size(); ++j) {
+        genomes.push_back(population.aSpecies[i].aGenomes[j]);
+      }
+    }
+  }
+
+  // Sort genomes by fitness. Smaller fitness is better and put first.
+  void SortGenomes(std::vector<Genome>& genomes) {
+    std::sort(genomes.begin(), genomes.end(), Species::CompareGenome);
+  }
+
+  // SortGenomes by adjusted fitness. Smaller is better.
+  static bool CompareGenomeByAdjustedFitness(Genome lg, Genome rg) {
+    return (lg.AdjustedFitness() < rg.AdjustedFitness());
+  }
+  void SortGenomesByAdjustedFitness(std::vector<Genome>& genomes) {
+    std::sort(genomes.begin(), genomes.end(), CompareGenomeByAdjustedFitness);
+  }
+
+  // Get genome index in a genomes vector.
+  ssize_t GetGenomeIndex(std::vector<Genome>& genomes, ssize_t id) {
+    for (ssize_t i=0; i<genomes.size(); ++i) {
+      if (genomes[i].Id() == id)
+        return i;
+    }
+    return -1;
+  }
+
+  // Calculate species' average rank in population by adjusted fitness. Bigger is better.
+  void CalcSpeciesAverageRank(Population& population, std::vector<double>& speciesAverageRank) {
+    std::vector<Genome> genomes;
+    SetAdjustedFitness(population);
+    AggregateGenomes(population, genomes);
+    SortGenomesByAdjustedFitness(genomes);  //!! we use adjusted fitness here for rank.
+    speciesAverageRank.clear();
+
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      double averageRank = 0;
+      ssize_t speciesSize = population.aSpecies[i].aGenomes.size();
+      for (ssize_t j=0; j<speciesSize; ++j) {
+        averageRank += genomes.size() - GetGenomeIndex(genomes, population.aSpecies[i].aGenomes[j].Id());
+      }
+      averageRank = averageRank / speciesSize;
+      speciesAverageRank.push_back(averageRank);
+    }
+  }
+
+  // Remove weak species.
+  void RemoveWeakSpecies(Population& population) {
+    std::vector<double> speciesAverageRank;
+    CalcSpeciesAverageRank(population, speciesAverageRank);
+    double totalAverageRank = std::accumulate(speciesAverageRank.begin(), speciesAverageRank.end(), 0);
+
+    for (ssize_t i=0; i<population.aSpecies.size(); ++i) {
+      double weak = (std::floor(speciesAverageRank[i] * population.NumSpecies() / totalAverageRank)
```

Comment:
I was thinking about the offspring calculation, but that's not what we are doing here, so my fault.